src/util.c

```diff
@@ -487,55 +489,88 @@ systemd_escape(char* string)
 gboolean
 netplan_delete_connection(const char* id, const char* rootdir)
 {
-    g_autofree gchar* del = NULL;
+    g_autofree gchar* yaml_path = NULL;
     g_autoptr(GError) error = NULL;
     NetplanNetDefinition* nd = NULL;
+    FILE* patch = NULL;
     gboolean ret = TRUE;
 
-    NetplanState* np_state = netplan_state_new();
-    NetplanParser* npp = netplan_parser_new();
+    NetplanParser* input_parser = netplan_parser_new();
+    NetplanState* input_state = netplan_state_new();
+    NetplanParser* output_parser = NULL;
+    NetplanState* output_state = NULL;
 
     /* parse all YAML files */
-    if (   !netplan_parser_load_yaml_hierarchy(npp, rootdir, &error)
-        || !netplan_state_import_parser_results(np_state, npp, &error)) {
+    if (   !netplan_parser_load_yaml_hierarchy(input_parser, rootdir, &error)
+        || !netplan_state_import_parser_results(input_state, input_parser, &error)) {
+        g_warning("netplan_delete_connection: Cannot parse input: %s\n", error->message);
+        ret = FALSE;
+        goto cleanup;
+    }
+
+    /* find specified netdef in input state */
+    nd = netplan_state_get_netdef(input_state, id);
+    if (!nd) {
+        g_fprintf(stderr, "netplan_delete_connection: Cannot delete %s, does not exist.", id);
+        ret = FALSE;
+        goto cleanup;
+    }
+
+    /* Build up a tab-separated YAML path for this Netdef (e.g. network.ethernets.eth0=...) */
+    yaml_path = g_strdup_printf("network\t%s\t%s", netplan_def_type_name(nd->type), id);
+
+    /* create a temporary file in memory, to hold our YAML patch */
+    patch = tmpfile();
+    int patch_fd = fileno(patch);
+    if (!netplan_util_create_yaml_patch(yaml_path, "NULL", patch_fd, &error)) {
         // LCOV_EXCL_START
-        g_fprintf(stderr, "%s\n", error->message);
+        g_fprintf(stderr, "netplan_delete_connection: Cannot create YAML patch: %s", error->message);
         ret = FALSE;
         goto cleanup;
         // LCOV_EXCL_STOP
     }
+    fflush(patch);
+
+    /* Create a new parser & state to hold our output YAML, ignoring the to be
+     * deleted Netdef from the patch */
+    output_parser = netplan_parser_new();
+    output_state = netplan_state_new();
 
-    if (!np_state->netdefs) {
+    rewind(patch);
+    if (   !netplan_parser_load_nullable_fields(output_parser, patch_fd, &error)
+        || !netplan_parser_load_yaml_hierarchy(output_parser, rootdir, &error)) {
         // LCOV_EXCL_START
```

Comment:
Just out of curiosity: why are we excluding those multiple error cases from coverage? Are they being tested in integration testing?

Reply:
Those are things which are really hard to test. They are not being tested in integration testing either. It's basically file system errors or errors in the emitter and/or parser of libyaml, which we cannot easily influence.

I guess I could mock some of the `netplan_` functions, when adding new unit tests as part of our new C based cmoka tests in `tests/ctest/`.. Maybe I should do this to test more of the error reporting cases.